### PR TITLE
Encrypt fetches pub key from auth

### LINKED ISSUE #264: Encrypt fetches pub key from auth
Updates needed:
 - [x] Endpoint implemented in neicnordic/sensitive-data-archive#200
 - [x] `sda-cli` to call it and initate the process as above
 This was already correctly implemented
 - [ ] there's a change in the config's fields for access_key to replace the `@` with `_`
 - [x] encrypt needs to look for the pub key directly in auth info endpoint as well. the current implementation only checks if it was put it .sda-cli-session config from a previous call on ./sda-cli login

### PR DESCRIPTION
**Pull request by pahatz**
closes #264 
Change encrypt module to fetch the public key from auth's endpoint info. Encrypt will now try to find a public key in the following order from:
- provided as an argument
- if `target` was provided, fetch it
- lookup in the .sda-cli-session from a previous login

### CODE REVIEW

File: encrypt/encrypt.go [hunk 1]
```diff
@@ -75,15 +82,37 @@ func Encrypt(args []string) error {
...
+			return err
+		}
+		log.Println("fetching public key")
+		// no key provided, no key in session file, target provided
+		publicKeyFileList = append(publicKeyFileList, pubFile)
+	}
+	// no key provided, no key in session file, no target provided
+	if len(publicKeyFileList) == 0 && sesKey == "" && *target == "" {
+		return errors.New("no public key could be obtained")
+	}
```
Comment (jbygdell) [inline encrypt/encrypt.go:115]: This entire block makes no logical sense.
 `GetPublicKeyFromSession` will return the name of the public key or an error.
That means that this block can be simplified to two if blocks:
```
if publicKeyFileList == nil && *target == "" {
...
```
and
```
if publicKeyFileList == nil && *target != "" {
...
```
or as a switch statement:
```
if publicKeyFileList == nil {
	switch len(*target) {
	case 0:
...
```

**Review by jbygdell — CHANGES_REQUESTED** [review]

**Comment by aaperis** [comment]
After discussions, the suggestion is that the `-target` flag should be mandatory so as to avoid reading info from the `.session` file that is for the wrong project.

**Comment by jbygdell** [comment]
> After discussions, the suggestion is that the `-target` flag should be mandatory so as to avoid reading info from the `.session` file that is for the wrong project.

That is not a good idea.
First of all, you can not prevent people from doing stupid things.
Second, it will make `sda-cli` unusable in airgaped/offline environments.

And if you look in the code the session file is stored in the current working directory, so the chance of a user having the "wrong" session file is slim, unless the user executes all commands from the exact same location.

The proper way to handle this would be to first sketch up the different scenarios for our 3 cases:
* specific public key - simple: if defined, use it.
* session file - last resort if neither `-key` nor `-target` is supplied
* `target` option - simple: if defined, use it.

Only one of `-key` or `-target` should be allowed.

**Comment by pahatz** [comment]
> > After discussions, the suggestion is that the `-target` flag should be mandatory so as to avoid reading info from the `.session` file that is for the wrong project.
> 
> That is not a good idea. First of all, you can not prevent people from doing stupid things. Second, it will make `sda-cli` unusable in airgaped/offline environments.
> 
> And if you look in the code the session file is stored in the current working directory, so the chance of a user having the "wrong" session file is slim, unless the user executes all commands from the exact same location.
> 
> The proper way to handle this would be to first sketch up the different scenarios for our 3 cases:
> 
>     * specific public key - simple: if defined, use it.
> 
>     * session file - last resort if neither `-key` nor `-target` is supplied
> 
>     * `target` option - simple: if defined, use it.
> 
> 
> Only one of `-key` or `-target` should be allowed.

I changed the order of the checks to respect this order now.

File: encrypt/encrypt.go [hunk 2]
```diff
@@ -26,7 +26,12 @@ import (
 // Usage text that will be displayed as command line help text when using the
 // `help encrypt` command
 var Usage = `
-USAGE: %s encrypt -key <public-key-file> (-outdir <dir>) (-continue=true) [file(s)]
+USAGE: %s encrypt -key <public-key-file> (-target <target>) (-outdir <dir>) (-continue=true) [file(s)]
+
+The target can be one of the following:
+	bp.nbis.se
+	fega.nbis.se
+	gdc.nbis.se
```
Comment (jbygdell) [inline encrypt/encrypt.go:34]:
```suggestion
```

remove all references to specific sites

File: encrypt/encrypt.go [hunk 3]
```diff
@@ -75,14 +82,41 @@ func Encrypt(args []string) error {
...
-	// no key provided, check for one in the session file
-	if len(publicKeyFileList) == 0 {
+	var pubKey string
 
-		sesKey, err := helpers.GetPublicKey()
+	if publicKeyFileList != nil && *target != "" {
+		return errors.New("only one of -key or -target can be used")
+	}
+
+	if publicKeyFileList == nil && *target != "" {
```
Comment (jbygdell) [inline encrypt/encrypt.go:91]:
```suggestion
	if *target != "" {
```

Since only one of them can be set at this point we only need to check the one we are interested in.

File: encrypt/encrypt.go [hunk 4]
```diff
@@ -75,14 +82,41 @@ func Encrypt(args []string) error {
...
+		if err != nil {
+			log.Println("could not read key from previous login,", err)
+		}
+		// key from session file found
+		if len(publicKeyFileList) == 0 && pubKey != "" {
+			publicKeyFileList = append(publicKeyFileList, pubKey)
+		}
+	}
+	if publicKeyFileList == nil {
+		return errors.New("no public key could be obtained")
```
Comment (jbygdell) [inline encrypt/encrypt.go:119]:
```suggestion
```
either `publicKeyFileList` has entries or we have returned an error by now.

File: encrypt/encrypt.go [hunk 5]
```diff
@@ -75,14 +82,41 @@ func Encrypt(args []string) error {
...
+		log.Println("fetching public key")
+		// no key provided, no key in session file, target provided
+		publicKeyFileList = append(publicKeyFileList, pubKeyFile)
+	}
+	// no key provided, no key in session file, no target provided
+	if publicKeyFileList == nil && *target == "" {
+		// check for public key in .sda-cli-session file from login
+		pubKey, err = helpers.GetPublicKeyFromSession()
+		if err != nil {
+			log.Println("could not read key from previous login,", err)
```
Comment (jbygdell) [inline encrypt/encrypt.go:111]:
```suggestion
			return err
```

File: encrypt/encrypt.go [hunk 6]
```diff
@@ -75,14 +82,41 @@ func Encrypt(args []string) error {
...
+		}
+		// key from session file found
+		if len(publicKeyFileList) == 0 && pubKey != "" {
+			publicKeyFileList = append(publicKeyFileList, pubKey)
+		}
```
Comment (jbygdell) [inline encrypt/encrypt.go:116]:
```suggestion
		publicKeyFileList = append(publicKeyFileList, pubKey)
```

File: helpers/helpers.go [hunk 7]
```diff
@@ -290,17 +290,16 @@ func GetAuth(path string) (*Config, error) {
...
 	}
 
-	if FileExists(".sda-cli-session") {
-		file, err := os.Open(".sda-cli-session")
-		if err != nil {
-			fmt.Println("could not read file:", file)
-		}
+	file, err := os.Open(".sda-cli-session")
+	if err != nil {
+		fmt.Println("could not read file:", file)
```
Comment (jbygdell) [inline helpers/helpers.go:302]:
```suggestion
		return "", err
```

File: helpers/helpers.go [hunk 8]
```diff
@@ -314,13 +313,24 @@ func GetPublicKey() (string, error) {
 		return "", errors.New("public key not found in the configuration")
 	}
 
+	pubFile, err := CreatePubFile(config.PublicKey, "key-from-oidc.pub.pem")
+	if err != nil {
+		return "", fmt.Errorf("failed to create public key file: %w", err)
```
Comment (jbygdell) [inline helpers/helpers.go:318]:
```suggestion
		return "", err
```

File: helpers/testcreate.pub.pem [hunk 9]
```diff
@@ -0,0 +1,3 @@
+-----BEGIN CRYPT4GH PUBLIC KEY-----
+MzM5ZWIyYTQ1OGZlYzVlMjNhYThiNTdjZmNiMzVmMTA=
+-----END CRYPT4GH PUBLIC KEY-----
```
Comment (jbygdell) [inline helpers/testcreate.pub.pem:3]: Don't commit this file

File: helpers/helpers_test.go [hunk 10]
```diff
@@ -370,10 +370,24 @@ public_key = 27be42445fd9e39c9be39e6b36a55e61e3801fc845f63781a813d3fe9977e17a
...
+	var expectedPubKey = `-----BEGIN CRYPT4GH PUBLIC KEY-----
+MzM5ZWIyYTQ1OGZlYzVlMjNhYThiNTdjZmNiMzVmMTA=
+-----END CRYPT4GH PUBLIC KEY-----
+`
+	_, err := CreatePubFile(pubKeyContent, "test_public_file.pub.pem")
+	assert.NoError(suite.T(), err)
+
+	pubFile, _ := os.ReadFile("test_public_file.pub.pem")
+	s := string(pubFile)
+	assert.Equal(suite.T(), expectedPubKey, s)
```
Comment (jbygdell) [inline helpers/helpers_test.go:392]: Here you need to remove the public key you wrote once the tests is completed.

File: login/login.go [hunk 11]
```diff
@@ -34,7 +34,7 @@ var ArgHelp = `
     [login-target]
         The login target can be one of the following: 
 			https://login.bp.nbis.se/
-			https://login.test.fega.nbis.se/
+			https://login.fega.nbis.se/
```
Comment (jbygdell) [inline login/login.go:37]: remove all references to specific sites.

File: login/login.go [hunk 12]
```diff
@@ -168,7 +168,7 @@ func NewLogin(args []string) error {
 	}
 	err = deviceLogin.Login()
 	if err != nil {
-		return fmt.Errorf("Login failed")
+		return fmt.Errorf("Login failed: %v", err)
```
Comment (jbygdell) [inline login/login.go:171]: Return the raw error and do the decoration in the calling function.

**Review by jbygdell — CHANGES_REQUESTED** [review]

File: encrypt/encrypt.go (hunk 2 above)
Comment (pahatz) [reply encrypt/encrypt.go:34]: There still should be some suggestion of how the target looks, or do you think that that information could be provided beforehand somewhere?

File: encrypt/encrypt.go (hunk 3 above)
Comment (pahatz) [reply encrypt/encrypt.go:87]: No, I think we still need that, otherwise we will get in this block in the case of having set the config as an argument.

File: encrypt/encrypt.go (hunk 2 above)
Comment (jbygdell) [reply encrypt/encrypt.go:34]: The target is an URL, we should not give explicit examples.

File: helpers/helpers_test.go [hunk 13]
```diff
@@ -389,3 +389,17 @@ func (suite *HelperTests) TestInvalidCharacters() {
...
 	}
 }
+
+func (suite *HelperTests) TestCreatePubFile() {
+	var pubKeyContent = `339eb2a458fec5e23aa8b57cfcb35f10e7389025816e44d4234f814ed2aeed3f`
+	var expectedPubKey = `-----BEGIN CRYPT4GH PUBLIC KEY-----
+MzM5ZWIyYTQ1OGZlYzVlMjNhYThiNTdjZmNiMzVmMTA=
+-----END CRYPT4GH PUBLIC KEY-----
+`
+	_, err := CreatePubFile(pubKeyContent, "test_public_file.pub.pem")
```
Comment (jbygdell) [inline helpers/helpers_test.go:399]: Put this in the standard temp directory instead of the current working directory

File: encrypt/encrypt.go (hunk 3 above)
Comment (jbygdell) [reply encrypt/encrypt.go:87]: Don't speculate.
If you are unsure, add a test case in the relevant test file.

File: encrypt/encrypt.go [hunk 14]
```diff
@@ -26,7 +26,8 @@ import (
 // Usage text that will be displayed as command line help text when using the
 // `help encrypt` command
 var Usage = `
-USAGE: %s encrypt -key <public-key-file> (-outdir <dir>) (-continue=true) [file(s)]
+USAGE: %s encrypt -key <public-key-file> (-target <target>) (-outdir <dir>) (-continue=true) [file(s)]
+
```
Comment (jbygdell) [inline encrypt/encrypt.go:30]:
```suggestion
```

**Review by jbygdell — CHANGES_REQUESTED** [review]

File: encrypt/encrypt.go (hunk 14 above)
Comment (pahatz) [reply encrypt/encrypt.go:30]: removed, not sure why github doesn't catch this

File: login/login.go (hunk 11 above)
Comment (pahatz) [reply login/login.go:37]: I fixed this even though it still shows up, remote is up to date

File: helpers/helpers_test.go (hunk 10 above)
Comment (pahatz) [reply helpers/helpers_test.go:404]: added

File: encrypt/encrypt.go (hunk 3 above)
Comment (pahatz) [reply encrypt/encrypt.go:86]: My phrasing was misleading.
The clause is correct, the tests actually caught it.

Comment (jbygdell) [reply encrypt/encrypt.go:86]: There are no relevant test cases in the code right now.
And since `publicKeyFileList` is supposed to be `nil` here we don't need to test for it, if it hadn't been nil it would have been caught on line 82.

File: encrypt/encrypt.go [hunk 15]
```diff
@@ -75,14 +77,38 @@ func Encrypt(args []string) error {
 		return err
 	}
 
-	// no key provided, check for one in the session file
-	if len(publicKeyFileList) == 0 {
+	var pubKey string
+
```
Comment (jbygdell) [inline encrypt/encrypt.go:81]:
```suggestion
```

This is not needed since the variable is only used inside the if block below

File: encrypt/encrypt.go [hunk 16]
```diff
@@ -75,14 +77,38 @@ func Encrypt(args []string) error {
...
+			return err
+		}
+		log.Println("fetching public key")
+		// no key provided, no key in session file, target provided
+		publicKeyFileList = append(publicKeyFileList, pubKeyFile)
+	}
+	// no key provided, no key in session file, no target provided
+	if publicKeyFileList == nil && *target == "" {
+		// check for public key in .sda-cli-session file from login
+		pubKey, err = helpers.GetPublicKeyFromSession()
```
Comment (jbygdell) [inline encrypt/encrypt.go:104]:
```suggestion
		pubKey, err := helpers.GetPublicKeyFromSession()
```

Do full declaration here

**Review by jbygdell — APPROVED** [review]
Other than a dangling variable declaration it now looks ok.

File: login/login.go (hunk 11 above)
Comment (pahatz) [reply login/login.go:37]: Is this coming from the two dot vs three dot diff that github does?
https://github.com/NBISweden/sda-cli/blob/264-fix-login-uris/login/login.go

Comment (jbygdell) [reply login/login.go:37]: No.
All URL needs to be removed.

File: login/login.go [hunk 17]
```diff
@@ -32,10 +32,7 @@ login:
 // the module help
 var ArgHelp = `
     [login-target]
-        The login target can be one of the following: 
-			https://login.bp.nbis.se/
-			https://login.test.fega.nbis.se/
-			https://login.gdi.nbis.se/`
+        The login target is the base URL of the service.`
```
Comment (jbygdell) [inline login/login.go:35]: Your previous wording was better

**Review by jbygdell — APPROVED** [review]

**Review by aaperis — APPROVED** [review]
Looks good!
 I think that a short description on the order of reading-in the pubkey should be added in the `Usage` text of `encrypt` so that the user is aware of the expected behavior. For example:
```
$ ./sda-cli encrypt README.md 
Error: configuration file (.sda-cli-session) not found
```
does not seem super user-friendly without any context.

But this is not crucial and neither is my other comment.

File: encrypt/encrypt.go [hunk 18]
```diff
@@ -75,14 +77,36 @@ func Encrypt(args []string) error {
...
 
-	// no key provided, check for one in the session file
-	if len(publicKeyFileList) == 0 {
+	if publicKeyFileList != nil && *target != "" {
+		return errors.New("only one of -key or -target can be used")
+	}
 
-		sesKey, err := helpers.GetPublicKey()
+	if *target != "" {
+		// fetch info endpoint values
```
[the remark this answers is not in the document]
Comment (pahatz) [reply encrypt/encrypt.go:85]: moved it

**Review by aaperis — APPROVED** [review]

**Review by jbygdell — APPROVED** [review]
